File: src/main/java/org/eigengo/activator/nashorn/NashornFuture.java
package org.eigengo.activator.nashorn;

import akka.japi.JavaPartialFunction;
import scala.Tuple2;
import scala.concurrent.ExecutionContext;
import scala.concurrent.Future;
import scala.concurrent.Future$;
import scala.util.Try;

import java.util.function.Function;

public class NashornFuture<A> {
    private final Future<A> future;

    public NashornFuture(Future<A> future) {
        this.future = future;
    }

    public static NashornFuture<?> failed(Throwable cause) {
        return new NashornFuture<>(Future$.MODULE$.failed(cause));
    }

    public <U> NashornFuture<U> flatMap(final Function<A, NashornFuture<U>> f, final ExecutionContext executor) {
        return new NashornFuture<>(this.future.flatMap(new JavaPartialFunction<A, Future<U>>() {
            @Override
            public Future<U> apply(A x, boolean isCheck) throws Exception {
                return f.apply(x).future;
            }
        }, executor));
    }

    public <U> NashornFuture<Tuple2<A, U>> zip(final NashornFuture<U> that) {
        return new NashornFuture<>(this.future.zip(that.future));
    }

    public void onComplete(final Function<Try<A>, ?> f, final ExecutionContext executor) {
        this.future.onComplete(new JavaPartialFunction<Try<A>, Object>() {
            @Override
            public Object apply(Try<A> x, boolean isCheck) throws Exception {
                f.apply(x);
                return null;
            }
        }, executor);
    }
}

File: src/main/scala/org/eigengo/activator/nashorn/japi/NashornFuture.scala
package org.eigengo.activator.nashorn.japi

import scala.concurrent.{ExecutionContext, Future}
import scala.util.{Failure, Success, Try}

object NashornFuture {

  def failed(t: Throwable): NashornFuture[_] = new NashornFuture(Future.failed(t))

  def fromScala[A](future: Future[A]): NashornFuture[A] = new NashornFuture(future)

}

class NashornFuture[A](private val future: Future[A]) {

  def flatMap[U](f: java.util.function.Function[A, NashornFuture[U]], executor: ExecutionContext): NashornFuture[U] =
    new NashornFuture(future.flatMap(x => f(x).future)(executor))

  def zip[U](that: NashornFuture[U]): NashornFuture[(A, U)] = new NashornFuture(this.future.zip(that.future))

  def onComplete(f: java.util.function.Function[Try[A], _], executor: ExecutionContext): Unit =
    future.onComplete(x => f(x))(executor)

  def onComplete2(s: java.util.function.Function[A, _], f: java.util.function.Function[Throwable, _], executor: ExecutionContext): Unit = {
    future.onComplete {
      case Success(x) => s(x)
      case Failure(t) => f(t)
    } (executor)
  }
}
